### ISSUE TITLE
no duration defined for given constraints

### ISSUE THREAD

**Issue by mturilli** [radical-cybertools/radical.analytics#37]
```
$ radical-stack 

  python               : 2.7.10
  pythonpath           : 
  virtualenv           : /Users/mturilli/Virtualenvs/RADICAL-ANALYTICS

  radical.analytics    : v0.45.2@devel
  radical.pilot        : split-71-gfdd36239@devel
  radical.utils        : v0.45-46-g591b33d@devel
  saga                 : split-14-g43c9ea2d@devel
```

With session `rp.session.titan-ext6.aleang9.017304.0005` I get the following error:
```
  File "/Users/mturilli/Virtualenvs/RADICAL-ANALYTICS/lib/python2.7/site-packages/radical/analytics/session.py", line 536, in duration
    ranges = self.ranges(state, event, time)
  File "/Users/mturilli/Virtualenvs/RADICAL-ANALYTICS/lib/python2.7/site-packages/radical/analytics/session.py", line 493, in ranges
    raise ValueError('no duration defined for given constraints')
ValueError: no duration defined for given constraints
```

Session at:

https://github.com/radical-experiments/rp-paper/tree/master/rawdata/exp1/gromacsweak-titan

**Comment by ibethune** [comment]
Adding to the pile of RA tickets to be looked at post-RP-0.46 release.

**Comment by mturilli** [comment]
Other example of failing session `rp.session.titan-ext2.aleang9.017309.0005` in `RADICAL-experiments/rp-paper` at `/analysis/data/use_case_id_5_gromacs/exp3/`. Same error, different stack:
```
$ radical-stack 

  python               : 2.7.10
  pythonpath           : 
  virtualenv           : /Users/mturilli/Virtualenvs/RADICAL-ANALYTICS

  radical.analytics    : v0.45.2@devel
  radical.pilot        : split-74-g10acb0d3@devel
  radical.utils        : v0.45-46-g591b33d@devel
  saga                 : v0.45.1@master
```

**Comment by andre-merzky** [comment]
@AA919 : the respecitve sessions seem to miss a final state for the pilots.  I am not exactly sure why no final state has been recorded - but one way to produce such profiles is to call `session.close(terminate=False)` or `session.close(delete=False)` (where `delete` is deprecated though). Is that used in your experiments, by any chance?

For the record: we now have a small script (again) which sets the final pilot profile to `CANCELED` at the time of the pmgr's demise, which should be a fair approximation for this specific set of experiments.

**Comment by AA919** [comment]
Hi,
I am sure that the only option in session.close is cleanup=False. Honestly, I did not even know the existence of that option. Tell me what I should look for and I will check. 

Cheers,

Alessio

**Comment by andre-merzky** [comment]
no worries Ale!  For completeness, the available options are:
* `cleanup`: remove the database (default is `True`)
* `terminate`: kill the pilots (default is `True`)
* `delete`: do both (deprecated)

Using `cleanup=False` should be safe - but then there must be another reason for the final state transition not being recorded.  Matteo now cleaned the session data, and we can use them all right, so please don't worry about that.  It looked like some runs had the final pilot state recorded all right, and some did not - which seems consistent with a bug rather then a wrong flag also.

I'll be sifting though the code to see if I can find an explanation.  My guess is that the pilot cancellation works and session termination the proceeds, but that the state callback is racing against the session termination at that point.  But, as said, that is just a guess, will have to look closer.

Bottom line: I'll close this analytics ticket, and open an RP ticket instead, if that's ok for you guys.

**Comment by andre-merzky** [comment]
see radical-cybertools/radical.pilot/issues/1362

**Comment by iparask** [comment]
I am getting this error with 0.47. What type of logs should I provide?

**Comment by andre-merzky** [comment]
What RA query are you running?  Are you sure that all units / pilots have the respective events / states?

**Comment by iparask** [comment]
This is my stack:

```
  python               : 2.7.13
  pythonpath           :
  virtualenv           : rp0.47

  radical.analytics    : 0.47.0
  radical.pilot        : 0.47.1
  radical.utils        : 0.47
  saga                 : 0.47
```

I am querying for CU `exec_start` and `exec_stop`. Out of 30 runs 17 of them produce this error. From those 17 I do not see any errors in the log files. The run was executed on Comet.

**Comment by andre-merzky** [comment]
Do you have `exec_start/exec_stop` for all units in the failing runs?  Can you please check?
```
found_start = False
found_stop = False
for unit in unit_session.get():
    for event in unit.events:
      if event[ru.EVENT] == 'exec_start' : found_start = True
      if event[ru.EVENT] == 'exec_stop'  : found_stop  = True
assert (found_start and found_stop)
```

or something like that...

**Comment by iparask** [comment]
Apparently, they do not have those events

```
In [1]: import pandas as pd
   ...: import radical.pilot as rp
   ...: import radical.analytics as ra
   ...: import radical.utils as ru
   ...: import glob
   ...: import os
   ...: import numpy as np
   ...:

In [2]: json_file = 'npConstRpMdaRmsd24_2/RpMdaRmsd24_2.json'

In [3]: src       = ''.join([x+'/' for x in json_file.split('/')[0:-1]])

In [4]: json      = ru.read_json(json_file)

In [5]: sid       = os.path.basename(json_file)[:-5]

In [6]: session   = ra.Session(sid=sid, stype='radical.pilot', src=src)

In [7]: unit_session = session.filter(etype='unit',inplace=False)

In [8]: found_start = False
   ...: found_stop = False
   ...: for unit in unit_session.get():
   ...:     for event in unit.events:
   ...:       if event[ru.EVENT] == 'exec_start' : found_start = True
   ...:       if event[ru.EVENT] == 'exec_stop'  : found_stop  = True
   ...: assert (found_start and found_stop)
   ...:
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-8-ae384290de6a> in <module>()
      5       if event[ru.EVENT] == 'exec_start' : found_start = True
      6       if event[ru.EVENT] == 'exec_stop'  : found_stop  = True
----> 7 assert (found_start and found_stop)

AssertionError:
```

Which means that I will open an RP ticket and close this one, since it is not an Analytics issue
